### PR TITLE
refactor: normalize variables resource

### PR DESCRIPTION
**Pull request by 121watts**
RFC: https://github.com/influxdata/idpe/issues/5591

### CODE REVIEW

File: ui/src/resources/reducers/helpers.ts [hunk 1]
```diff
@@ -46,12 +39,9 @@ export const editResource = <R>(
   resource: ResourceType
 ) => {
   const {entities, result} = action.schema
-  if (!draftState.byID) {
-    draftState.byID = {}
-  }
 
-  const bucket = entities[resource][result]
-  draftState.byID[result] = bucket
+  const r = entities[resource][result]
```
Comment (hoorayimhelping) [inline ui/src/resources/reducers/helpers.ts]: why have a transient variable called `r`? `r` doesn't provide any more meaning than `entities[resource][result]` what do you think about changing this to:

```ts
draftState.byID[result] = entities[resource][result]
```
?

File: ui/src/schemas/index.ts [hunk 2]
```diff
@@ -85,9 +95,30 @@ export const arrayOfTelegrafs = [telegraf]
...
+    processStrategy: (v: Variable): Variable => {
+      return {
+        ...v,
+        labels: addLabels(v),
+        status: addStatus(v),
+      }
+    },
   }
+)
+export const arrayOfVariables = [variable]
```
Comment (hoorayimhelping) [inline ui/src/schemas/index.ts]: what's this all about? doing a `cmd+f` for it doesn't show it used in this pr

File: ui/src/types/schemas.ts [hunk 3]
```diff
@@ -66,3 +65,11 @@ export interface TaskEntities {
     [uuid: string]: Task
   }
 }
+
+// VariableEntities defines the result of normalizr's normalization
```
Comment (hoorayimhelping) [inline ui/src/types/schemas.ts]: 👍

File: ui/src/types/variables.ts [hunk 4]
```diff
@@ -1,7 +1,63 @@
...
+
+export interface ValueSelections {
+  [variableID: string]: string
+}
+
+export interface VariablesState extends NormalizedState<Variable> {
+  values: {
+    // Different variable values can be selected in different
+    // "contexts"---different parts of the app like a particular dashboard, or
+    // the Data Explorer
```
Comment (hoorayimhelping) [inline ui/src/types/variables.ts]: love this comment

File: ui/src/variables/actions/thunks.ts [hunk 5]
```diff
@@ -365,51 +280,57 @@ export const addVariableLabelAsync = (
...
+      schemas.variable
+    )
 
     dispatch(setVariable(variableID, RemoteDataState.Done, variable))
   } catch (error) {
     console.error(error)
-    dispatch(notify(removeVariableLabelFailed()))
+    dispatch(notify(copy.removeVariableLabelFailed()))
   }
 }
```
Comment (hoorayimhelping) [inline ui/src/variables/actions/thunks.ts]: love the changes in this file, great work!

File: ui/src/variables/reducers/editor.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+})
+
+export const variableEditorReducer = (
+  state: VariableEditorState = initialEditorState(),
+  action: EditorAction
+): VariableEditorState =>
+  produce(state, draftState => {
+    switch (action.type) {
+      case CLEAR_VARIABLE_EDITOR: {
+        return initialEditorState()
```
Comment (hoorayimhelping) [inline ui/src/variables/reducers/editor.ts]: does `draftState` need to be set to `initialEditorState` here? this case is kinda different in form from all the other cases.

File: ui/src/variables/selectors/index.tsx [hunk 7]
```diff
@@ -15,29 +15,28 @@ import {
...
+type VariablesState = ResourceState['variables']['byID']
+type ValuesState = ResourceState['variables']['values']['contextID']
 
 const extractVariablesListMemoized = memoizeOne(
   (variablesState: VariablesState): Variable[] => {
-    return Object.values(variablesState)
-      .filter(d => d.status === RemoteDataState.Done)
-      .map(d => d.variable)
+    return Object.values(variablesState).filter(
+      d => d.status === RemoteDataState.Done
```
Comment (hoorayimhelping) [inline ui/src/variables/selectors/index.tsx]: what is `d`?

Comment (121watts) [reply ui/src/variables/selectors/index.tsx]: idk.  it was there  when i got there.

File: ui/src/variables/reducers/editor.ts (hunk 6 above)
Comment (121watts) [reply ui/src/variables/reducers/editor.ts]: nope! https://immerjs.github.io/immer/docs/return

File: ui/src/schemas/index.ts (hunk 2 above)
Comment (121watts) [reply ui/src/schemas/index.ts]: Should normalizer want to normalize a list of a foo it uses this syntax for it's schema.  I could just inline this, but (imo), `variable` and `[variable]` looked to similar when consumed as arguments to the `normalize` function i.e.:

`normalize(data, fooSchema)` and `normalize(data, [fooSchema])` looked really similar but return different results.  So, to make it easier on the eye I assigned all resource list schemas to a variable.

File: ui/src/variables/selectors/index.tsx (hunk 7 above)
Comment (121watts) [reply ui/src/variables/selectors/index.tsx]: idk `d` was there when i got there.  😄

File: ui/src/variables/reducers/editor.ts (hunk 6 above)
Comment (121watts) [reply ui/src/variables/reducers/editor.ts]: nope!  https://immerjs.github.io/immer/docs/return

File: ui/src/resources/reducers/helpers.ts (hunk 1 above)
Comment (121watts) [reply ui/src/resources/reducers/helpers.ts]: its certainly better than `buckets`.  it's really a resource but resource is already taken 🤷‍♂.  Good enough for golang!

File: ui/src/schemas/index.ts (hunk 2 above)
Comment (121watts) [reply ui/src/schemas/index.ts]: I can remove it if it's not used.

File: ui/src/variables/selectors/index.tsx (hunk 7 above)
Comment (hoorayimhelping) [reply ui/src/variables/selectors/index.tsx]: this is how i feel about little things like this:

[>don't ya see it? well pick it up!!](https://www.youtube.com/watch?v=wwQ5SmhwGgk&feature=youtu.be&t=156)

basically it's all our job to sweep the floor and we should keep the place tidy for the next person. anything we can do to reduce our collective mental load. to that point, maybe `data` is slightly better than `d`?

File: ui/src/resources/reducers/helpers.ts (hunk 1 above)
Comment (hoorayimhelping) [reply ui/src/resources/reducers/helpers.ts]: the point i'm making is that it's already clear from the assignment and the name `entities[resource][result]` that `r` represents a result, and if we can eliminate one more line of code without losing any meaning, i'm generally 110% for that.

>Good enough for golang!

we're not writing go though, we're writing typescript / javascript

**Pull request merged**